server/Arcadia.Assistant.CSP/Vacations/DatabaseVacationsCache.cs

```diff
@@ -0,0 +1,121 @@
+﻿namespace Arcadia.Assistant.CSP.Vacations
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using Arcadia.Assistant.Calendar.Abstractions;
+
+    internal class DatabaseVacationsCache
+    {
+        private readonly Dictionary<string, CalendarEventWithApprovals> cache;
+
+        public DatabaseVacationsCache(Dictionary<string, CalendarEventWithApprovals> values)
+        {
+            this.cache = new Dictionary<string, CalendarEventWithApprovals>(values);
+        }
+
+        public CalendarEventWithApprovals this[string index]
+        {
+            get => this.cache[index];
+            set => this.cache[index] = value;
+        }
+
+        public Diff Difference(Dictionary<string, CalendarEventWithApprovals> values)
+        {
+            var createdEvents = new List<CalendarEventWithApprovals>();
+            var updatedEvents = new List<CalendarEventWithApprovals>();
+            var approvalsUpdatedEvents = new List<CalendarEventWithApprovals>();
+
+            foreach (var @event in values.Values)
+            {
+                if (!this.cache.ContainsKey(@event.CalendarEvent.EventId))
+                {
+                    if (@event.CalendarEvent.Status == VacationStatuses.Requested)
+                    {
+                        if (!@event.Approvals.Any())
+                        {
+                            createdEvents.Add(@event);
+                        }
+                        else
+                        {
+                            approvalsUpdatedEvents.Add(@event);
+                        }
+                    }
+                    else
+                    {
+                        updatedEvents.Add(@event);
+                    }
+                }
+                else
+                {
+                    var cacheEvent = this.cache[@event.CalendarEvent.EventId];
```

Comment:
TryGetValue(out var cacheEvent) instead of ContainsKey

Reply:
Ok, but what's real profit?

Reply:
Done.